pootle/apps/pootle_translationproject/views.py

```diff
@@ -50,6 +50,51 @@ def admin_permissions(request, translation_project):
                        'translation_projects/admin/permissions.html', ctx)
 
 
+def get_sidebar_announcements_context(request, project, language):
+    announcements = []
+    new_cookie_data = {}
+    cookie_data = {}
+
+    if SIDEBAR_COOKIE_NAME in request.COOKIES:
+        json_str = unquote(request.COOKIES[SIDEBAR_COOKIE_NAME])
+        cookie_data = json.loads(json_str)
+
+    is_sidebar_open = cookie_data.get('isOpen', True)
+
+    def get_announcement(virtual_path):
+        try:
+            announcement = StaticPage.objects.live(request.user).get(
+                virtual_path=virtual_path,
+            )
+        except StaticPage.DoesNotExist:
+            pass
+        else:
+            announcements.append(announcement)
+
+            ann_mtime = dateformat.format(announcement.modified_on, 'U')
+            stored_mtime = cookie_data.get(virtual_path, None)
+
+            if ann_mtime != stored_mtime:
+                is_sidebar_open = True
+                new_cookie_data[virtual_path] = ann_mtime
+
+    get_announcement('announcements/projects/' + project.code)
```

Comment:
It looks like you misunderstood or my explanation wasn't clear enough: the idea I wanted highlight was to keep things easy to reason about and not have functions that have many side-effects (single responsibility principle).

With adjusting `get_announcement()` I meant something like:

```python
def get_announcement(project_code=None, language_code=None):
    if language_code is None and project_code is None:
        return None

    if language_code is None and project_code is not None:
        path = u'/projects/%s' % (project_code,)
    else:
        path = u'/'.join(filter(None, [language_code, project_code]))

    virtual_path = u'/announcements/%s/' % (path, )
    try:
        return StaticPage.objects.live(request.user).get(
            virtual_path=virtual_path,
        )
    except StaticPage.DoesNotExist:
        return None    
```

`get_announcement()` does one thing and does it well (the path construction could even be factored out into an utility function), and callers don't need to know about internal implementation details such as what a `virtual_path` is and how it's constructed specifically for announcements.